### PR TITLE
LPS-172418 Relationship Tab is not exported properly

### PR DESCRIPTION
**Pull request by seltong**
Related issue: [LPS-172418](https://issues.liferay.com/browse/LPS-172418)

### CODE REVIEW

**Comment by marcelabc** [comment]
Adding the Label "Blocked" because I'm waiting for a decision from BPM leadership regarding adding the ERC in the Object Relationship entity.

**Comment by seltong** [comment]
ci:test:sf

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectDefinitionLocalServiceTest.java [hunk 1]
```diff
@@ -1330,6 +1283,53 @@ public void testEnableAccountEntryRestrictedForNondefaultStorageType()
 		_objectDefinitionLocalService.deleteObjectDefinition(objectDefinition);
 	}
 
+	@Test
+	public void testEnableAccountRestricted() throws Exception {
+		ObjectDefinition objectDefinition =
```
Comment (seltong) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectDefinitionLocalServiceTest.java:1288]: @seltong Check this

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java [hunk 2]
```diff
@@ -854,18 +855,18 @@ private ObjectRelationship _addObjectRelationship(
 			boolean reverse, String type)
 		throws PortalException {
 
+		User user = _userLocalService.getUser(userId);
+
 		_validate(
-			externalReferenceCode, objectDefinitionId1, objectDefinitionId2,
-			parameterObjectFieldId, name, type);
+			externalReferenceCode, user.getCompanyId(), objectDefinitionId1,
```
Comment (seltong) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java:861]: Check this

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:test:object

**Comment by seltong** [comment]
ci:test:relevant

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:test:relevant

File: modules/apps/object/object-service/service.xml [hunk 3]
```diff
@@ -556,6 +560,11 @@
 			<finder-column name="objectDefinitionId2" />
 			<finder-column name="reverse" />
 		</finder>
+		<finder name="ERC_C_ODI1" return-type="ObjectRelationship">
```
Comment (marcelabc) [inline modules/apps/object/object-service/service.xml:563]: I saw some examples with this additional property `unique="true"` 
Could you investigate if we should add this, please?

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java [hunk 4]
```diff
@@ -561,6 +561,14 @@ public ObjectRelationship getObjectRelationship(
 		}
 	}
 
+	@Override
+	public ObjectRelationship getObjectRelationshipByExternalReferenceCode(
```
Comment (marcelabc) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java:568]: I think we should rename this method for `fetchObjectRelationshipByExternalReferenceCode()` if we call the method `fetch` in the persistence class.

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java [hunk 5]
```diff
@@ -1013,11 +1014,21 @@ private ObjectRelationship _updateObjectRelationship(
...
+		if (Validator.isNotNull(externalReferenceCode)) {
+			ObjectRelationship objectRelationship =
+				objectRelationshipPersistence.fetchByERC_C_ODI1(
+					externalReferenceCode, companyId, objectDefinitionId1);
+
+			if (objectRelationship != null) {
+				throw new DuplicateObjectRelationshipExternalReferenceCodeException();
+			}
+		}
+
```
Comment (marcelabc) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java:1031]: Could you create a separated method only to validate the ERC, please?
Like: _validateExternalReferenceCode()
Murilo refactored this validate method and separated the validation about the properties `name` and `type` (https://github.com/liferay-objects/liferay-portal/pull/2322);

Comment (marcelabc) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java:1031]: When creating this method, could you please add the failure first like this:
```
if (Validator.isNull(externalReferenceCode)) {
	return;
}
```

File: modules/apps/object/object-admin-rest-impl/src/main/java/com/liferay/object/admin/rest/internal/resource/v1_0/ObjectLayoutResourceImpl.java [hunk 6]
```diff
@@ -292,7 +294,28 @@ private com.liferay.object.model.ObjectLayoutRow _toObjectLayoutRow(
...
 	private com.liferay.object.model.ObjectLayoutTab _toObjectLayoutTab(
-		long objectDefinitionId, ObjectLayoutTab objectLayoutTab) {
+			long objectDefinitionId, ObjectLayoutTab objectLayoutTab)
+		throws PortalException {
+
+		long objectRelationshipId = GetterUtil.getLong(
+			objectLayoutTab.getObjectRelationshipId());
+
+		if (objectLayoutTab.getObjectRelationshipExternalReferenceCode() !=
+				null) {
```
Comment (marcelabc) [inline modules/apps/object/object-admin-rest-impl/src/main/java/com/liferay/object/admin/rest/internal/resource/v1_0/ObjectLayoutResourceImpl.java:304]: When it is a String, we should validate using this method `Validator.isNull()`.
When it is an Object, we should use !=.

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java [hunk 7]
```diff
@@ -105,14 +105,16 @@ public class ObjectRelationshipLocalServiceImpl
 	@Indexable(type = IndexableType.REINDEX)
 	@Override
 	public ObjectRelationship addObjectRelationship(
-			long userId, long objectDefinitionId1, long objectDefinitionId2,
```
Comment (marcelabc) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java:108]: I believe you only added this parameter in the ADD methods. I believe we have to add this parameter and validate it in the update methods.

Comment (marcelabc) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java:108]: I will check with Gabriel/Camacho if we will allow the edition of the ERC in the Object Relationship. But, I believe that we have to allow to maintain the pattern.

**Review by marcelabc — CHANGES_REQUESTED** [review]
Hi 
I finished my review. Let me know if you have questions :smile:

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:test:relevant

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java [hunk 8]
```diff
@@ -695,8 +706,9 @@ public void registerObjectRelationshipsRelatedInfoCollectionProviders(
 	@Indexable(type = IndexableType.REINDEX)
 	@Override
 	public ObjectRelationship updateObjectRelationship(
-			long objectRelationshipId, long parameterObjectFieldId,
-			String deletionType, Map<Locale, String> labelMap)
+			String externalReferenceCode, long objectRelationshipId,
```
Comment (marcelabc) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectRelationshipLocalServiceImpl.java:709]: As we discuss, add externalReferenceCode property in  _updateObjectRelationship()

**Review by marcelabc — CHANGES_REQUESTED** [review]
Hi 
I finished my review. Let me know if you have questions :smile:

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:test:object

**Review by marcelabc — CHANGES_REQUESTED** [review]
As we discussed, @seltong will add a "validation" to the object relationship id in the ObjectLayoutTabLocalServiceImpl class

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:test:object

**Comment by seltong** [comment]
ci:test:relevant

**Review by marcelabc — APPROVED** [review]
LGTM 🎉

**Comment by marcelabc** [comment]
Rebase errors @seltong  :confused:

**Comment by marcelabc** [comment]
Hey @gabrielwas,
If you want to check this PR about ERC in Object Relationship :smile:

**Comment by marcelabc** [comment]
Hey @seltong,
After resolving the rebase error, could you:
- compile, check the integration tests again
- run the CI
- ask for the QA team to test locally in their Slack channel, please. I already talked with André so, when you do this request, you can tag him and me please. When you ask, could you give a little of context of this task and why do we need to test locally (upgrade/import export/ ERC ...), please?

Thank you!
Let me know if you have questions :smile:

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:test:relevant

**Comment by seltong** [comment]
Resolving conflicts...

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:test:relevant

**Review by gabrielwas — CHANGES_REQUESTED** [review]
Can we increment the integration tests to cover this implementation? A simple check is enough.

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:test:relevant

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
ci:test:relevant

**Comment by seltong** [comment]
ci:test:object

**Comment by seltong** [comment]
ci:test:sf

**Comment by timpak** [comment]
ci:test:bundle

**Pull request closed without merging**